asr1k_neutron_l3/plugins/l3/service_plugins/l3_extension_adapter.py

```diff
@@ -317,8 +320,41 @@ def get_sync_data(self, context, router_ids=None, active=None, host=None):
             router["rt_export"] = list(set(rt_export))
             router["rt_import"] = list(set(rt_import))
 
+            all_ports = [x["id"] for x in router.get("_interfaces", [])]
+            if gw_port:
+                all_ports.append(gw_port["id"])
+
+            router["fwaas_policies"] = self.get_fwaas_policies(context, all_ports)
+
         return routers
 
+    def get_fwaas_policies(self, context, port_ids):
+        policies = {}
+
+        for port_id in port_ids:
+            with context.session.begin(subtransactions=True):
+                fwg_id = self.db.get_fwg_attached_to_port(context, port_id)
+                if not fwg_id:
+                    continue
+                fwg = self.db.get_firewall_group(context, fwg_id,
+                                                 fields=["egress_firewall_policy_id",
+                                                        "ingress_firewall_policy_id",
+                                                        "name", "id"])
+                for fwp_id, direction in ((fwg["egress_firewall_policy_id"], "egress"),
+                                          (fwg["ingress_firewall_policy_id"], "ingress")):
+                    if fwp_id is None:
+                        continue
+                    if fwp_id not in policies:
+                        name = self.db.get_firewall_policy(context, fwp_id, fields=["name"])["name"]
+                        policies[fwp_id] = {
+                            "name": name,
+                            "ingress_ports": [],
+                            "egress_ports": [],
+                            "rules": self.db._get_policy_ordered_rules(context, fwp_id)
+                        }
+                    policies[fwp_id][f"{direction}_ports"].append(port_id)
```

Comment:
It does something similar. It returns the rules that are transitively connected (via a policy) to a firewall group. Yet in our FWaaS implementation we are pushing firewall policies as ACLs on the box. So for us it's important to get the UUID of the policy and the ports and direction the policy is attached. 

We could use `make_firewall_group_dict_with_rules`, but that would entail creating a ACL per firewall group, which would also work, it would just lead to us creating more ACLs with potentially the same content. ACLs and ACEs are limited resources (even though IIRC on the Cat 8500 the limit is so high that this wouldn't create a problem), so I chose the approach requiring less ACLs, ACEs and thus less config changes.